runtime/src/block_component_processor.rs

```diff
@@ -194,6 +200,35 @@ impl BlockComponentProcessor {
         Ok(())
     }
 
+    fn verify_genesis_certificate(
+        bank: &Bank,
+        cert: &Certificate,
+    ) -> Result<(), BlockComponentProcessorError> {
+        let slot = cert.cert_type.slot();
+        let epoch = bank.epoch_schedule().get_epoch(slot);
+        let epoch_stakes = bank
+            .epoch_stakes_map()
+            .get(&epoch)
+            .ok_or(BlockComponentProcessorError::GenesisCertificateFailedVerification)?;
+        let key_to_rank_map = epoch_stakes.bls_pubkey_to_rank_map();
+        let total_stake = epoch_stakes.total_stake();
+
+        let genesis_stake =
+            cert_verify::verify_votor_message_certificate(cert, key_to_rank_map.len(), |rank| {
+                key_to_rank_map
+                    .get_pubkey_and_stake(rank)
+                    .map(|(_, bls_pubkey, stake)| (*bls_pubkey, *stake))
+            })
+            .map_err(|_| BlockComponentProcessorError::GenesisCertificateFailedVerification)?;
+
+        let genesis_percent = Fraction::new(genesis_stake, NonZeroU64::new(total_stake).unwrap());
```

Comment:
should we just check `total_stake` is non-zero and return error here?

Reply:
I did think about it, but if `total_stake` is `0` we're already screwed. Probably `epoch_stakes.total_stake()` should be returning a `NonZero`.